Sources/SwiftDocC/Infrastructure/Extensions/DeclarationFragments+Simplify.swift

```diff
@@ -0,0 +1,78 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2024 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import SymbolKit
+
+fileprivate typealias Fragment = SymbolKit.SymbolGraph.Symbol.DeclarationFragments.Fragment
+
+extension [Fragment] {
+    /// Simplifies a declaration into its overloaded representation by removing type idenfiers.
+    ///
+    /// Consider the following Swift code:
+    ///
+    /// ```swift
+    /// class MyClass {
+    ///     func myFunc(param: Int) -> Int {}
+    ///     func myFunc(param: String) -> String {}
+    /// }
+    /// ```
+    ///
+    /// These two methods can both be referred to as `myFunc(param:)`, regardless of their
+    /// parameter and return types. However, when generating an overload group's landing page, we
+    /// copy the symbol information from one of these methods, including the type names in the
+    /// sub-heading and navigator titles. Ideally, we would want to be able to display
+    /// `func myFunc(param:)` in curation and navigation to prevent any confusion.
+    ///
+    /// The idea behind this function is that any overload will have the same basic display
+    /// fragments other than any type names, and that (in Swift) these type names (and only these
+    /// type names) are all rendered as `typeIdentifier` fragments. Therefore, we can strip out
+    /// these references and get a unified representation.
+    ///
+    /// However, doing this by itself will return `func myFunc(param: ) -> ` as the result. After
+    /// stripping out any `typeIdentifier` fragments, this method then cleans up any visual
+    /// oddities like the space in the parameters list and the trailing return-type arrow.
+    func simplifyForOverloads() -> Self {
```

Comment:
Please add dedicated tests for this function with more complex example data. This current  implementation doesn't work well for array types, optional types, inout parameters, borrowed/consuming parameters, tuples, closures, variadics, parameter packs, and more.

For example, I defined this function
```swift
public func someFunction<Result: Equatable>(
    first: borrowing [Int?],
    second: inout Int,
    third: (Int, Int),
    fourth: @autoclosure () throws -> Void,
    fifth: Int...,
    sixth: some Sequence<Int>,
    seventh: any Hashable
) -> Result? {
    fatalError()
}
```

which gets this sub heading in the symbol graph file (the full declaration has even more tokens)

```
func someFunction<Result>(first: borrowing [Int?], second: inout Int, third: (Int, Int), fourth: () throws -> Void, fifth: Int..., sixth: some Sequence<Int>, seventh: any Hashable) -> Result?
```

Calling `simplifyForOverloads()` on that sub heading results in

```
func someFunction<Result>(first: borrowing [?], second: inout , third: (,), fourth: () throws, fifth: ..., sixth: some <>, seventh: any)?
```

which still contains a lot of tokens which I would expect to be filtered out.

----

Similarly, a function definition like

```swift
public func somePackedFunction<each T: Collection>(
    _ item: repeat each T
) -> (repeat (each T).Element?) {
    fatalError()
}
```

has this sub heading

```
func somePackedFunction<each T>(repeat each T) -> (repeat (each T).Element?)
```

and calling `simplifyForOverloads()` on it results in 

```
simplified  'func somePackedFunction<each T>(repeat each) -> (repeat (each).?)'
```

which also contains a number of tokens that I would expect to be filtered out.

Reply:
Maybe it would be worth building on top / adding tokens to the page title as opposed to stripping out unwanted fragments from the subheading/navigator, since it seems like you want a slightly longer form of the title? I don't know how often new tokens are added (maybe rarely), but that could reduce the number of unexpected tokens that might slip through here if new tokens are added to subheadings in the future.

Reply:
@d-ronnqvist I think there are alternate filterings and techniques i can try to deal with the situations you've mentioned. Thanks for the examples; i'll work up some tests for these.

@mayaepps The issue with starting from the title is that the title is plain text whereas the navigator and subheading are based on declaration fragments. This is important since it affects how Swift-DocC-Render will color a curated symbol link, for example with this test page i made with the current implementation:

![image](https://github.com/apple/swift-docc/assets/5217170/c81aa4ec-3dfd-4248-91ea-26f26b4e0b0d)

I believe it's worth keeping as much rich information as we can, to preserve this behavior.

Reply:
You could possible build up the simplified subheading based on the parameters names (and external parameter names) in the type signature.

Reply:
Right, i feel like we could get away with filtering out everything but `keyword`, `identifier`, `externalParam`, and `text` fragments, and then stripping out everything that comes after the return-type arrow.